ostd/src/mm/page_table/cursor.rs

```diff
@@ -692,51 +691,47 @@ where
         assert!(end <= self.0.barrier_va.end);
 
         while self.0.va < end {
-            let cur_pte = self.0.read_cur_pte();
-            if !cur_pte.is_present() {
+            let cur_va = self.0.va;
+            let cur_level = self.0.level;
+            let cur_is_tracked = self.0.should_map_as_tracked();
+            let mut cur_entry = self.0.cur_entry();
+
+            // Skip if it is already absent.
+            if cur_entry.is_none() {
                 self.0.move_forward();
                 continue;
             }
 
-            // Go down if it's not a last node.
-            if !cur_pte.is_last(self.0.level) {
-                self.0.level_down();
-
-                // We have got down a level. If there's no mapped PTEs in
-                // the current node, we can go back and skip to save time.
-                if self.0.guards[(self.0.level - 1) as usize]
-                    .as_ref()
-                    .unwrap()
-                    .nr_children()
-                    == 0
-                {
-                    self.0.level_up();
+            // Go down if it's not a last entry.
+            if cur_entry.is_node() {
+                let Child::PageTable(pt) = cur_entry.to_owned() else {
+                    unreachable!("Already checked");
+                };
+                let pt = pt.lock();
+                // If there's no mapped PTEs in the next level, we can
+                // skip to save time.
+                if pt.nr_children() != 0 {
+                    self.0.push_level(pt);
+                } else {
                     self.0.move_forward();
                 }
-
                 continue;
             }
 
             // Go down if the page size is too big and we are protecting part
             // of untracked huge pages.
-            if self.0.va % page_size::<C>(self.0.level) != 0
-                || self.0.va + page_size::<C>(self.0.level) > end
-            {
-                if self.0.in_tracked_range() {
-                    panic!("protecting part of a huge page");
-                } else {
-                    self.level_down_split();
-                    continue;
-                }
+            if cur_va % page_size::<C>(cur_level) != 0 || cur_va + page_size::<C>(cur_level) > end {
+                debug_assert!(cur_entry.is_last());
+                assert!(!cur_is_tracked, "Protecting part of a huge page");
+                let split_child = cur_entry.split_untracked_huge();
```

Comment:
The above two `debug_assert!`/`assert!` becomes duplicate with the assertion in `split_untrack_huge`. Can we make it a `split_if_untracked_huge` and just unwrap here?